core-plugins/src/main/java/co/cask/hydrator/plugin/batch/sink/KVTableSink.java

```diff
@@ -91,9 +129,40 @@ public KVTableSink(KVTableConfig kvTableConfig) {
   public void transform(StructuredRecord input, Emitter<KeyValue<byte[], byte[]>> emitter) throws Exception {
     Object key = input.get(kvTableConfig.keyField);
     Preconditions.checkArgument(key != null, "Key cannot be null.");
-    byte[] keyBytes = key instanceof ByteBuffer ? Bytes.toBytes((ByteBuffer) key) : (byte[]) key;
+    byte[] keyBytes;
+
+    Schema keyFieldSchema = input.getSchema().getField(kvTableConfig.keyField).getSchema();
+    if (keyFieldSchema.getType().equals(Schema.Type.STRING)) {
+      keyBytes = Bytes.toBytes((String) key);
+    } else if (keyFieldSchema.getType().equals(Schema.Type.BYTES)) {
+      keyBytes = key instanceof ByteBuffer ? Bytes.toBytes((ByteBuffer) key) : (byte[]) key;
+    } else if (keyFieldSchema.isNullable()) {
+      throw new Exception("Key is nullable in the input record : " + input);
+    } else {
+      throw new Exception(
+        "Key is of unsupported type in the record" + input + ", only String and Bytes are supported");
```

Comment:
don't log the record, as StructuredRecord has not overridden toString() as anything useful. Message should be something like "Key field %s cannot have schema %s. It must be a bytes or string.".